File: Editor/Tools/UI/CreateTogglePanelWithButtonTool.cs
using System;
using UnityEngine;
using UnityEngine.UI;
using UnityEditor;
using McpUnity.Utils;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;
using TMPro;

namespace McpUnity.Tools
{
    public class CreateTogglePanelWithButtonTool : McpToolBase
    {
        public CreateTogglePanelWithButtonTool()
        {
            Name = "create_toggle_panel_with_button";
            Description = "Creates a dark gray semi-transparent panel with a button that toggles its visibility.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                // Parse parameters
                string panelName = parameters["panelName"]?.ToObject<string>() ?? "TogglePanel";
                string buttonText = parameters["buttonText"]?.ToObject<string>() ?? "Toggle Panel";
                float panelPosX = parameters["panelPosX"]?.ToObject<float>() ?? 0f;
                float panelPosY = parameters["panelPosY"]?.ToObject<float>() ?? 0f;
                float panelWidth = parameters["panelWidth"]?.ToObject<float>() ?? 400f;
                float panelHeight = parameters["panelHeight"]?.ToObject<float>() ?? 300f;
                string panelColorHex = parameters["panelColor"]?.ToObject<string>() ?? "#333333D9";
                float buttonPosX = parameters["buttonPosX"]?.ToObject<float>() ?? 0f;
                float buttonPosY = parameters["buttonPosY"]?.ToObject<float>() ?? 200f;
                float buttonWidth = parameters["buttonWidth"]?.ToObject<float>() ?? 160f;
                float buttonHeight = parameters["buttonHeight"]?.ToObject<float>() ?? 40f;
                bool panelInitiallyActive = parameters["panelInitiallyActive"]?.ToObject<bool>() ?? true;

                // Find or create Canvas
                Canvas canvas = UnityEngine.Object.FindFirstObjectByType<Canvas>();
                if (canvas == null)
                {
                    GameObject canvasObj = new GameObject("Canvas");
                    canvas = canvasObj.AddComponent<Canvas>();
                    canvas.renderMode = RenderMode.ScreenSpaceOverlay;
                    canvasObj.AddComponent<CanvasScaler>();
                    canvasObj.AddComponent<GraphicRaycaster>();
                    
                    // Ensure EventSystem exists
                    if (UnityEngine.Object.FindFirstObjectByType<UnityEngine.EventSystems.EventSystem>() == null)
                    {
                        GameObject eventSystemObj = new GameObject("EventSystem");
                        eventSystemObj.AddComponent<UnityEngine.EventSystems.EventSystem>();
                        eventSystemObj.AddComponent<UnityEngine.EventSystems.StandaloneInputModule>();
                        Undo.RegisterCreatedObjectUndo(eventSystemObj, "Create EventSystem");
                    }
                    
                    Undo.RegisterCreatedObjectUndo(canvasObj, "Create Canvas");
                }

                // Create Panel
                GameObject panelObj = new GameObject(panelName);
                panelObj.transform.SetParent(canvas.transform, false);

                RectTransform panelRect = panelObj.AddComponent<RectTransform>();
                panelRect.anchoredPosition = new Vector2(panelPosX, panelPosY);
                panelRect.sizeDelta = new Vector2(panelWidth, panelHeight);

                Image panelImage = panelObj.AddComponent<Image>();
                
                // Parse panel color (default to dark gray with 85% opacity)
                if (ColorUtility.TryParseHtmlString(panelColorHex, out Color panelColor))
                {
                    panelImage.color = panelColor;
                }
                else
                {
                    panelImage.color = new Color(0.2f, 0.2f, 0.2f, 0.85f); // Dark gray, 85% opacity
                }

                // Set initial active state
                panelObj.SetActive(panelInitiallyActive);

                Undo.RegisterCreatedObjectUndo(panelObj, "Create Toggle Panel");

                // Create Button
                GameObject buttonObj = new GameObject(buttonText);
                buttonObj.transform.SetParent(canvas.transform, false);
                
                RectTransform buttonRect = buttonObj.AddComponent<RectTransform>();
                buttonRect.anchoredPosition = new Vector2(buttonPosX, buttonPosY);
                buttonRect.sizeDelta = new Vector2(buttonWidth, buttonHeight);

                Image buttonImage = buttonObj.AddComponent<Image>();
                buttonImage.color = Color.white;

                Button button = buttonObj.AddComponent<Button>();

                // Wire button to toggle panel
                button.onClick.AddListener(() => {
                    panelObj.SetActive(!panelObj.activeSelf);
                });

                // Create Text child for button
                GameObject textObj = new GameObject("Text");
                textObj.transform.SetParent(buttonObj.transform, false);
                
                RectTransform textRect = textObj.AddComponent<RectTransform>();
                textRect.anchorMin = Vector2.zero;
                textRect.anchorMax = Vector2.one;
                textRect.sizeDelta = Vector2.zero;

                // Try to use TextMeshPro, fallback to standard Text
                try
                {
                    TextMeshProUGUI tmpText = textObj.AddComponent<TextMeshProUGUI>();
                    tmpText.text = buttonText;
                    tmpText.alignment = TextAlignmentOptions.Center;
                    tmpText.color = Color.black;
                }
                catch
                {
                    Text text = textObj.AddComponent<Text>();
                    text.text = buttonText;
                    text.alignment = TextAnchor.MiddleCenter;
                    text.color = Color.black;
                    text.font = UnityEngine.Resources.GetBuiltinResource<Font>("Arial.ttf");
                }

                Undo.RegisterCreatedObjectUndo(buttonObj, "Create Toggle Button");

                return new JObject
                {
                    ["success"] = true,
                    ["message"] = $"Created toggle panel '{panelName}' with button '{buttonText}'.",
                    ["panelName"] = panelName,
                    ["panelInstanceId"] = panelObj.GetInstanceID(),
                    ["buttonText"] = buttonText,
                    ["buttonInstanceId"] = buttonObj.GetInstanceID(),
                    ["panelPosition"] = $"({panelPosX}, {panelPosY})",
                    ["panelSize"] = $"({panelWidth}, {panelHeight})",
                    ["buttonPosition"] = $"({buttonPosX}, {buttonPosY})",
                    ["buttonSize"] = $"({buttonWidth}, {buttonHeight})",
                    ["panelInitiallyActive"] = panelInitiallyActive
                };
            }
            catch (Exception ex)
            {
                McpLogger.LogError($"CreateTogglePanelWithButtonTool error: {ex.Message}");
                return McpUnitySocketHandler.CreateErrorResponse($"Failed: {ex.Message}", "execution_error");
            }
        }
    }
}

File: Server~/src/tools/ui/CreateTogglePanelWithButtonTool.ts
import { z } from 'zod';
import { BaseTool } from '../base/BaseTool.js';
import { Tool, Tags } from '../base/ToolDecorators.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

/**
 * Input schema for create_toggle_panel_with_button tool
 */
const CreateTogglePanelWithButtonToolArgsSchema = z.object({
  panelName: z.string().optional().default('TogglePanel').describe('Name for the Panel GameObject. Default: "TogglePanel"'),
  buttonText: z.string().optional().default('Toggle Panel').describe('Text to display on the toggle button. Default: "Toggle Panel"'),
  panelPosX: z.number().optional().default(0).describe('X position of panel in canvas space. Default: 0'),
  panelPosY: z.number().optional().default(0).describe('Y position of panel in canvas space. Default: 0'),
  panelWidth: z.number().optional().default(400).describe('Width of the panel. Default: 400'),
  panelHeight: z.number().optional().default(300).describe('Height of the panel. Default: 300'),
  panelColor: z.string().optional().default('#333333D9').describe('Panel color in hex format with alpha. Default: "#333333D9" (dark gray, 85% opacity)'),
  buttonPosX: z.number().optional().default(0).describe('X position of button in canvas space. Default: 0'),
  buttonPosY: z.number().optional().default(200).describe('Y position of button in canvas space. Default: 200'),
  buttonWidth: z.number().optional().default(160).describe('Width of the button. Default: 160'),
  buttonHeight: z.number().optional().default(40).describe('Height of the button. Default: 40'),
  panelInitiallyActive: z.boolean().optional().default(true).describe('Whether the panel should be initially visible. Default: true')
});

/**
 * CreateTogglePanelWithButton Tool
 * Creates a dark gray semi-transparent panel with a button that toggles its visibility
 */
@Tool({
  name: 'create_toggle_panel_with_button',
  description: 'Creates a dark gray semi-transparent panel with a button that toggles its visibility',
  category: 'ui',
  version: '1.0.0'
})
@Tags(['unity', 'ui', 'panel', 'button', 'toggle'])
export class CreateTogglePanelWithButtonTool extends BaseTool {
  get name() {
    return 'create_toggle_panel_with_button';
  }

  get description() {
    return 'Creates a dark gray semi-transparent panel with a button that toggles its visibility';
  }

  get inputSchema() {
    return CreateTogglePanelWithButtonToolArgsSchema;
  }

  get category() {
    return 'ui';
  }

  protected formatSuccessResponse(result: any): CallToolResult {
    return {
      content: [{
        type: 'text',
        text: `✅ ${result.message || 'Operation completed successfully'}\n` +
              `Panel: ${result.panelName} (ID: ${result.panelInstanceId})\n` +
              `Button: ${result.buttonText} (ID: ${result.buttonInstanceId})\n` +
              `Click the button to toggle panel visibility.`
      }]
    };
  }
}

